CIP-0022/CIP-0022.md

```diff
@@ -0,0 +1,113 @@
+
+pooltool.io - Validates operators by receiving 1 ada spent from the pool's registered rewards account
+
+adapools.org - Validates operators by requesting that the operator include a special generated value in their extended pool metadata json file.
+
+This proposal desires to simplify and streamline a single approach that all can reference in order to verify that a pool operator is who they say they are.
+
+Current Tools that have implemented it already or have plans to implement it:
+* [CNCLI](https://github.com/AndrewWestberg/cncli)
+* [JorManager](https://bitbucket.org/muamw10/jormanager/)
+* [StakePoolOperator Scripts](https://github.com/gitmachtl/scripts)
+* [CNTools](https://cardano-community.github.io/guild-operators/#/Scripts/cntools)
+
+
+# Specification
+
+In order to achieve the goals of this CIP, the pool operator needs to provide some credential or credentials to the validating party which cannot be spoofed. The VRF pool keys and VRF signature algorithm implemented in libsodium are chosen to build and provide this credential/signature. This signature can then be validated and the operator verified without ever exposing any of the pool's private key information. This technique is very similar to verifying that a block produced by another pool is valid. The only difference is that instead of validating the slot seed for a given pool, we're validating a pre-determined message hash.
+
+Code Example (Pool Operator side):
+``` 
+// Node operational VRF-Verification-Key: pool.vrf.vkey
+//{
+//    "type": "VrfVerificationKey_PraosVRF",
+//    "description": "VRF Verification Key",
+//    "cborHex": "5820e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381"
+//}
+//
+// Node operational VRF-Signing-Key: pool.vrf.skey
+//{
+//    "type": "VrfSigningKey_PraosVRF",
+//    "description": "VRF Signing Key",
+//    "cborHex": "5840adb9c97bec60189aa90d01d113e3ef405f03477d82a94f81da926c90cd46a374e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381"
+//}
+
+// We assume the pool operator has access to the pool's vrf secret key
+val skeyCbor = "5840adb9c97bec60189aa90d01d113e3ef405f03477d82a94f81da926c90cd46a374e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381".hexToByteArray()
+val vrfSkey = (CborReader.createFromByteArray(skeyCbor).readDataItem() as CborByteString).byteArrayValue()
+val vkeyCbor = "5820e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381".hexToByteArray()
+val vrfVkey = (CborReader.createFromByteArray(vkeyCbor).readDataItem() as CborByteString).byteArrayValue()
+
+// Client side, sign the hash of a pre-determined message
+val message = "pooltool.io".toByteArray()
```

Comment:
I'd suggest for the message have some pre-defined structure. It's generally not a good idea to encourage people signing arbitrary messages with their secret credentials. For example:

```rs
val message = "cip-0022{pooltool.io}".toByteArray()
```

Would already prevent unfortunate scenarios where some adversarial online service tries to forge a VRF signature by providing some well-crafted message to the operator. The signing code of the operator should always encapsulate the message into its structure and not blindly sign whatever is sent to him/her.

Reply:
The challenge _should_ / _must_ also contains a random nonce to prevent replayability (I could send you a message that some other online tool would send you and re-use your signature to hijack your identity on that other service, supposing you haven't already registered there).

Reply:
We can add this, but we're already mutating the input before it's signed. We are actually signing the blake2b hash of the input.

Reply:
@AndrewWestberg true that, but most messages that are signed are actually hashed before. Transactions, blocks etc... Some signing algorithms don't even support messages that are longer than a certain length anyway.

Reply:
about the signing, we really should have functions in the cardano-cli that allow users to sign and verify "content" via shelley keys. like its possible with jormungander(jcli). this feature is missing in cardano-cli totally.

Reply:
@KtorZ Definitely need to add some wording about the Verification server's challenge including a random nonce that they send. That's a very good point about a replay attack. I'll re-work this CIP to incorporate your ideas.

Reply:
@KtorZ Please take a look at the updated CIP. I hope I've integrated all of your suggestions correctly.